pkg/crypto/pedersen/points.go

```diff
@@ -2049,5 +2048,76 @@ func init() {
 		y, _ := new(big.Int).SetString(p[1], 16)
 		points[i] = point{x, y}
 	}
-	curve = weierstrass.Stark()
+	P, _ = new(big.Int).SetString("800000000000011000000000000000000000000000000000000000000000001", 16)
+	A = big.NewInt(1)
+}
+
+// Add returns the sum of (x1, y1) and (x2, y2) on an
+// elliptic curve mod P.
+// See https://github.com/starkware-libs/cairo-lang/blob/2abd303e1808612b724bc1412b2b5babd04bb4e7/src/starkware/python/math_utils.py#L147-L164
+func (p *point) Add(p2 point) {
+	if p.x.Sign() == 0 && p.y.Sign() == 0 {
+		p.x.Set(p2.x)
+		p.y.Set(p2.y)
+		return
+	}
+
+	if p2.x.Sign() == 0 && p2.y.Sign() == 0 {
+		return
+	}
+
+	x1, y1 := new(big.Int).Mod(p.x, P), new(big.Int).Mod(p.y, P)
+	x2, y2 := new(big.Int).Mod(p2.x, P), new(big.Int).Mod(p2.y, P)
+	if x1.Cmp(x2) == 0 {
+		temp := new(big.Int).Sub(P, y2)
+		temp.Mod(temp, P)
+		if y1.Cmp(temp) == 0 {
+			p.x.SetInt64(0)
+			p.y.SetInt64(0)
+		} else {
+			big2 := big.NewInt(2)
+
+			p.x = new(big.Int).Mul(x1, x1)
+			p.x.Mul(p.x, big.NewInt(3))
+			p.x.Add(p.x, A)
+			m := divMod(p.x, new(big.Int).Mul(y1, big2), P)
+
+			p.x = new(big.Int).Mul(m, m)
+			p.x.Sub(p.x, new(big.Int).Mul(big2, x1))
+			p.x.Mod(p.x, P)
+
+			p.y = new(big.Int).Sub(x1, p.x)
+			p.y.Mul(p.y, m)
+			p.y.Sub(p.y, y1)
+			p.y.Mod(p.y, P)
+		}
+	} else {
+		x1, y1 := new(big.Int).Mod(p.x, P), new(big.Int).Mod(p.y, P)
+		x2, y2 := new(big.Int).Mod(p2.x, P), new(big.Int).Mod(p2.y, P)
+
+		xDelta := new(big.Int).Sub(x1, x2)
+		yDelta := new(big.Int).Sub(y1, y2)
+
+		m := divMod(yDelta, xDelta, P)
+
+		p.x = new(big.Int).Mul(m, m)
+		p.x.Sub(p.x, x1)
+		p.x.Sub(p.x, x2)
+		p.x.Mod(p.x, P)
+
+		p.y = new(big.Int).Sub(x1, p.x)
+		p.y.Mul(p.y, m)
+		p.y.Sub(p.y, y1)
+		p.y.Mod(p.y, P)
+	}
```

Comment:
Lines 2087-2092 and 2106-2111 are exactly the same. It seems to me that they could be factored out both if blocks.